### PR TITLE
Advanced TSC 3 & PEONS & Hackinator & Bugfixes

### PR DESCRIPTION
**Pull request by Leo40Git**
This allows each parameter to have it's own length.
It also fixes two glaring issues that I have no idea how I missed last time:

- Every TSC list got the "basic" treatment, which caused errors with advanced lists
- The "clearsMsg" attribute didn't actually do anything

This'll probably be the last "advanced TSC" pull request.

### CODE REVIEW

**Comment by Leo40Git** [comment]
This is pretty important, as without it the whole "advanced TSC parsing" thing doesn't actually work.

**Comment by Leo40Git** [comment]
DON'T MERGE THIS YET!!! 20kdc and I are trying to figure out why BL and PEONS together break horribly, and how to fix it.

**Comment by Leo40Git** [comment]
...okay 20kdc just gave up on fixing it. Great.
Expect possible activity from txin tomorrow. Maybe.

**Comment by Leo40Git** [comment]
So initializing a new segment using PEONS and having BL initialize it worked...
Unfortunately, attempting to add a new segment on top of that did _not_ work...

**Comment by Leo40Git** [comment]
Hackinator stuff, including the image field type Enlight wanted and a default value attribute.
Default value might not work and image field type might break horribly.

**Comment by Leo40Git** [comment]
okay this pull request has basically turned into a "general changes" pull request
sorry about that

File: src/ca/noxid/lab/gameinfo/GameInfo.java [hunk 1]
```diff
@@ -88,17 +88,19 @@ public void patch(ByteBuffer data, int offset) {
 	public GameInfo(File base) throws IOException {
 		mapdataStore = new Vector<>();
 		categoryMap = new HashMap<>();
+		File configFile = new File(dataDir + "/bl.ini"); //$NON-NLS-1$
+		gameConfig = new BlConfig(configFile, type);
 		if (base.toString().endsWith(".exe")) { //$NON-NLS-1$
 			dataDir = new File(base.getParent() + "/data"); //$NON-NLS-1$
-			
-			if (dataDir.list(new FileSuffixFilter("stprj")).length > 0) {
+			// dummy out GUXT """support"""
```
Comment (taedixon) [inline src/ca/noxid/lab/gameinfo/GameInfo.java]: :^)

Comment (Leo40Git) [reply src/ca/noxid/lab/gameinfo/GameInfo.java]: Is that even close to being functional?

File: tsc_list.txt [hunk 2]
```diff
@@ -37,94 +37,94 @@ Format:
...
-<TUR	0	----	Text UnRead?	Instantly display text. Use after a <MSG/2/3; works until another <MSG/2/3 or an <END.	0	0	1	4
-<UNI	1	#---	UNIverse?	Set character movement type. Use 0000 for normal, 0001 for zero-G and 0002 to disallow movement.	0	0	1	4
-<UNJ	2	#e--	UNiverse Jump	Jump to event X if movement is of type W (0000 for normal, 0001 for zero-G).	0	0	1	4
-<WAI	1	.---	WAIt	Pause script for W ticks.	0	0	1	4
-<WAS	0	----	WAit until Standing	Pause script until character is on ground.	0	0	1	4
-<XX1	1	l---	XX1	Show the island falling in manner W. Use 0000 to have it crash and 0001 to have it stop midway.	0	0	1	4
-<YNJ	1	e---	Yes/No Jump	Prompt Yes/No; jump to event W if No is selected.	0	0	1	4
-<ZAM	0	----	Zero ArMs	Sets all weapon energy to zero.	0	0	1	4
\ No newline at end of file
+<AE+	0	----	Arms Energy +	Refill all weapon ammo.	0	0	1	4	4	4	4
```
Comment (taedixon) [inline tsc_list.txt]: this file is getting kind of wild.. I guess it serves its purpose but maybe at some point it would be nice if there was an easier way to maintain it.

Comment (Leo40Git) [reply tsc_list.txt]: I'm working on a TSC list editor.

**Comment by taedixon** [comment]
this PR got a bit out of hand, I guess it's my fault for not getting around to the review sooner lol
to clarify, most of the changes to the hackinator are code formatting? or is there something important I should be looking for in there? it seems to take up most of the diff

File: src/ca/noxid/lab/gameinfo/GameInfo.java [hunk 3]
```diff
@@ -26,7 +26,7 @@
 	private String imageExtension;
 	private File dataDir;
 	private BlConfig gameConfig;
-	private static final String CSPLUS_IMG_EXT = ".png";
+	private static final String CSPLUS_IMG_EXT = ".bmp";
```
Comment (taedixon) [inline src/ca/noxid/lab/gameinfo/GameInfo.java]: you're right on assuming this might break some custom stuff but at this point it's so old as to be irrelevant anyway. I really thought it used PNGs... my mind must be playing tricks on me.

**Comment by Leo40Git** [comment]
Hackinator diffs are the `image` field type which (hopefully) allows displaying images in the hackinator,
and the `defvalue` attribute which (hopefully) allows hacks to have default values in textboxes and such.

**Comment by Leo40Git** [comment]
Important lines in Hackinator are:
https://github.com/taedixon/boosters-lab/pull/18/files#diff-40399566b88f0154188f2fffa533d7f5R388
https://github.com/taedixon/boosters-lab/pull/18/files#diff-40399566b88f0154188f2fffa533d7f5R409
https://github.com/taedixon/boosters-lab/pull/18/files#diff-40399566b88f0154188f2fffa533d7f5R580

**Comment by Leo40Git** [comment]
Note to self: After this gets merged, add an issue about translating (and actually using!) messages_jp.properties... not in this pull request, though. This one is already loaded to the brim.

**Comment by Leo40Git** [comment]
last version may or may not have fucked .csmap up if it was smaller than 95 maps, regardless if it was actually empty or not
also i accidentally formatted the code again, sorry

**Comment by Leo40Git** [comment]
@20kdc has stopped giving up. Unfortunately, he isn't making much progress either.
~~IF ONLY WINDOWS GAVE SOME MORE INFO ON WHY IT CAN'T RUN AN EXE~~

**Comment by taedixon** [comment]
is this uh... are you done now? should I start reviewing?

**Comment by Leo40Git** [comment]
Done now. You can start reviewing.

File: src/ca/noxid/lab/mapdata/MapInfo.java [hunk 4]
```diff
@@ -921,7 +921,7 @@ public byte getTileB(int x, int y, int layer) {
 	public int getTile(int x, int y, int layer) {
 		int rVal = 0;
 		try {
-			if (EditorApp.EDITOR_MODE == 0) {
+			if (EditorApp.EDITOR_MODE == 0 && layer == -1) {
```
Comment (taedixon) [inline src/ca/noxid/lab/mapdata/MapInfo.java]: not sure why this change was necessary? in what situations will layer be -1?
should the && have been an || maybe?

File: src/ca/noxid/lab/rsrc/ResourceManager.java [hunk 5]
```diff
@@ -183,7 +183,7 @@ private BufferedImage black2Trans(BufferedImage src)
 			for (int x = 0; x < src.getWidth(); x++)
 			{
 				int px = src.getRGB(x, y);
-				if (px == -16777216) //argb black full opaque
+				if (px == Color.BLACK.getRGB())
```
Comment (taedixon) [inline src/ca/noxid/lab/rsrc/ResourceManager.java]: good change tbh

File: src/ca/noxid/lab/tile/MapPane.java [hunk 6]
```diff
@@ -445,15 +446,20 @@ protected void drawTileTypes(Graphics2D g) {
...
 						tileType = dataHolder.calcPxa(tile1);
 					} else {
 						tileType = dataHolder.calcPxa(tile2);
+						tileLayer = 2;
 					}
 				} else {
 					tileType = dataHolder.calcPxa(tile2);
+					tileLayer = 2;
 				}
+				if (!visArray[tileLayer])
```
Comment (taedixon) [inline src/ca/noxid/lab/tile/MapPane.java]: not so sure on this change. for one, i'd imagine that one could want to not see the "physical" tiles of the tile layer while still showing the tile types and you can achieve the same effect of hiding both by unticking the "show tile types". 
secondly, if you're gonna skip rendering the whole layer it should be done outside the loop to save performance.

File: src/ca/noxid/lab/gameinfo/GameInfo.java [hunk 7]
```diff
@@ -90,15 +90,24 @@ public GameInfo(File base) throws IOException {
 		categoryMap = new HashMap<>();
 		if (base.toString().endsWith(".exe")) { //$NON-NLS-1$
 			dataDir = new File(base.getParent() + "/data"); //$NON-NLS-1$
-			
+			File configFile = new File(dataDir + "/bl.ini"); //$NON-NLS-1$
+			gameConfig = new BlConfig(configFile, type);
```
[the remark this answers is not in the document]
Comment (taedixon) [reply src/ca/noxid/lab/gameinfo/GameInfo.java]: i pushed a change to master to fix another related issue with this

File: src/ca/noxid/lab/mapdata/MapInfo.java (hunk 4 above)
Comment (Leo40Git) [reply src/ca/noxid/lab/mapdata/MapInfo.java]: Adding that seems to fix the layer visibility settings (and the equivalent addition in putTile seems to fix undoing/redoing).

**Comment by Leo40Git** [comment]
okay so an idea on how to fix the .excode thing (dummied out for now):
since resizing the mapdata seems to be the only breaking point, i could make it so that adding an .excode segment will also lock the mapdata segment to a certain size (and thus disable adding and removing maps but who actually uses that feature?)
this will also require a small change to the mapdata segment format: 4 bytes (integer) at the start of the segment that tells BL how many maps are actually there
your opinion?

**Comment by Leo40Git** [comment]
Actually, this *does* allow removing and adding maps, it just doesn't allow resizing the mapdata (and puts a limit on the map count but you find me a person that uses more than 1024 maps).

**Comment by Leo40Git** [comment]
**I'M DONE, I'M DONE.**
No more commits. Unless you tell me to change something in a review

**Comment by taedixon** [comment]
your PR includes Java 1.8 features however the project language level is 1.7

**Comment by Leo40Git** [comment]
Too used to lambdas... Care to point out where I use Java 8 features, please?

**Comment by taedixon** [comment]
wouldn't it be easiest to set whatever compilation tool you're using to language level 7?
there's a few in CSExe, `sectionsSorted.sort`, `Integer.parseUnsignedInt`, 
maybe more the IDE only reports the first one it finds to me

**Comment by Leo40Git** [comment]
Eclipse only applies the compiler level to features (lambdas, <>, etc.), *not* to methods and fields. I'd probably have to download JDK7 for that.
